pwndbg/heap/ptmalloc.py

```diff
@@ -547,11 +487,906 @@ def largebin_index(self, sz):
+                # try to find `mstate ar_ptr = &main_arena;` in malloc_trim instructions
+                malloc_trim_instructions = pwndbg.disasm.near(pwndbg.symbol.address('malloc_trim'), 10,
+                                                              show_prev_insns=False)
+                if pwndbg.arch.current == "x86-64":
+                    for instr in malloc_trim_instructions:
+                        # try to find `lea rax,[rip+DISP]`
+                        if instr.mnemonic == 'lea' and "rip" in instr.op_str and instr.disp > 0:
+                            self._main_arena_addr = instr.next + instr.disp # rip + disp
+                            break
+                elif pwndbg.arch.current == "i386":
+                    base_offset = pwndbg.vmmap.find(pwndbg.symbol.address('_IO_list_all')).start
+                    for instr in malloc_trim_instructions:
+                        # try to find `lea edi,[eax+DISP]`
+                        if instr.mnemonic == 'lea' and "eax" in instr.op_str and instr.disp > 0:
+                            self._main_arena_addr = base_offset + instr.disp # eax + disp
+                            break
+                elif pwndbg.arch.current == "aarch64":
+                    base_offset = pwndbg.vmmap.find(pwndbg.symbol.address('_IO_list_all')).start
+                    reg = None
+                    for instr in malloc_trim_instructions[5:]:
+                        # Try to find `add reg2, reg1, #offset` after `adrp reg1, #base_offset`
+                        if instr.mnemonic == 'add' and instr.operands[1].str == reg:
+                            self._main_arena_addr = base_offset + instr.operands[2].int
+                            break
+                        if instr.mnemonic == 'adrp' and instr.operands[1].int == base_offset:
+                            reg = instr.operands[0].str
+                elif pwndbg.arch.current == "arm":
+                    ldrw_instr = None
+                    for instr in malloc_trim_instructions:
+                        # Try to find `ldr.w reg, [pc, #offset]`, then `add reg, pc`
+                        if not ldrw_instr:
+                            if instr.mnemonic == 'ldr.w':
+                                ldrw_instr = instr
+                        else:
+                            reg = ldrw_instr.operands[0].str
+                            if instr.mnemonic == 'add' and instr.op_str == reg + ", pc":
+                                # ldr.w reg, [pc, #offset]
+                                offset = ldrw_instr.operands[1].mem.disp
+                                offset = pwndbg.memory.s32((ldrw_instr.address + 4 & -4) + offset)
+                                # add reg, pc
+                                self._main_arena_addr = offset + instr.address + 4
+                else:
+                    inform_report_issue("main_arena")
+                    raise OSError("Cannot find the symbol via heuristics")
```

Comment:
Does this make sense? We have the last resort check of finding `main_arena` via _IO_* symbols in libc in lines 742+ --- shall we really raise an exception here instead of trying out a different approach?

Reply:
We have an "else" with the same code at the end of the whole function, so maybe we should not fallback here and try more methods

Reply:
Make sense!
Thanks, I'll fix it.